### ISSUE TITLE
AB test is touching Apache

### ISSUE THREAD

**Issue by ADDISON74** [nexcess/magento-turpentine#1273]
I just installed the latest devel branch in a fresh Ubuntu Sever 16.04 for testing. Everything works as expected if I load Magento frontend in browser:

- ESI blocks are in source code
- Page load time is in a range of 2ms  - 7 ms
- Age of the files is increasing with every load
- Disabling caching is proving the files are coming from Varnish
- Cookies are set as usual for frontend

I am doing an AB test from other machine to check the number of requests. In the past with a test like ab -n1000 -c1000 http://www.mydomain.com/ I could get 2000 - 3000 request/second. Now I am getting 65 requests/seconds. Checking the memory and instances of Apache there is one conclusion ab test is touching directly the Apache server. When Varnish is serving the files normally the memory should remain on the same position, no CPU load, this is what I get in the past. Now a basic test like n100 c10 did for 10 times is blocking my system.

Normally Turpentine should detect AB request and serve them, but it doesn't make that. Did you make changes in how cookies are processed? If I did something wrong please correct me. I did this too many times in the past getting impressive results in the number of requests using Varnish in front of Apache.

**Comment by miguelbalparda** [comment]
Can you try running siege with --header="Cookie: frontend=niu8v9qubdm1q0a17t7qb7dn82"?

**Comment by ADDISON74** [comment]
<strong>First test without cookie setup</strong>:

``` siege www.mydomain.com/category1 -r10 -c50 -d1```

Results:
```
** SIEGE 3.0.8
** Preparing 50 concurrent users for battle.
The server is now under siege..      done.

Transactions:                    500 hits
Availability:                 100.00 %
Elapsed time:                  20.97 secs
Data transferred:               4.03 MB
Response time:                  1.15 secs
Transaction rate:              23.84 trans/sec
Throughput:                     0.19 MB/sec
Concurrency:                   27.48
Successful transactions:         500
Failed transactions:               0
Longest transaction:           11.41
Shortest transaction:           0.00
```

<strong>Second test as you requested</strong>:

``` siege www.mydomain.com/category1 -r10 -c50 -d1 --header="Cookie: frontend=niu8v9qubdm1q0a17t7qb7dn82"```

Results:
```
** SIEGE 3.0.8
** Preparing 50 concurrent users for battle.
The server is now under siege..      done.

Transactions:                    500 hits
Availability:                 100.00 %
Elapsed time:                  15.43 secs
Data transferred:               1.25 MB
Response time:                  0.50 secs
Transaction rate:              32.40 trans/sec
Throughput:                     0.08 MB/sec
Concurrency:                   16.25
Successful transactions:           0
Failed transactions:               0
Longest transaction:            4.98
Shortest transaction:           0.00
```

As you can see the number of requests is pretty the same like touching Apache directly. But the frontend is loading fast, page load is 2 ms - 7 ms. If someone will do a siege or ab test on my server it will blow up for a while.

**Comment by miguelbalparda** [comment]
Yeah the difference is not enormous but there are other variables like response time and throughput being cut in half. 
I'm also curious about Successful transactions: 0. Thoughts?
Can you try to do the same with a valid session key from your store?

**Comment by ADDISON74** [comment]
<strong>Observation 1</strong>: From time to time I get no_external_cookie =1. I try to delete it but is coming back. This is happening even "Disable caching" option in Chrome Inspect ->Network  is on or off. After deleting all sessions in Magento folder and clearing the browser cache I can get rid off this cookie.

<strong>Observation 2</strong>:  Keeping F5 pressed in browser without Varnish active is flooding the webserver. CPU and memory are going nuts. Having Varnish active with Turpentine working and keeping again F5, CPU is a little bit used, memory stays on the same level. This proves Varnish + Apache work.

<strong>Observation 3</strong>: Apache Logs. When Varnish service is stopped and leave only Apache working the webserver log is recording Apache Benchmarks as it should do. When Varnish service is working the log is recording only GET turpentine/esi/getBlock/method/esi ..." for every request.

<strong>Observation 4</strong>: A siege test with a valid session key

```siege www.mydomain.com/category1 -r10 -c50 -d1 --header="Cookie: frontend=ndfviohbr7odbrbqfh467dre00"```

Results:

```
** SIEGE 3.0.8
** Preparing 50 concurrent users for battle.
The server is now under siege..      done. 
Transactions:                     500 hits
Availability:                 100.00 %
Elapsed time:                  22.84 secs
Data transferred:               8.57 MB
Response time:                  1.25 secs
Transaction rate:              21.89 trans/sec
Throughput:                     0.38 MB/sec
Concurrency:                   27.26
Successful transactions:         500
Failed transactions:               0
Longest transaction:           13.04
Shortest transaction:           0.04
```

I don't understand why the number of requests is still so low. I remember I could load so easy any page on the webserver in the past when I was running AB test in parallel from other computer. Now it is dead with AB and siege in testing I have to reboot to get back to normal.

**Comment by ADDISON74** [comment]
What a strange situation. In Admin -> System -> Cache Management only [Varnish Pages] and [Varnish ESI Blocks] options were enable. I disabled them both then restarted Varnish service. There are no ESI tags in the page source code but first AB test provided very good results. From now on every AB test is providing an average of 1800 requests/second. In htop CPU is not blinking and nothing in plus to the memory.  (CPU 2.27 GHz x 8 threads, 16 GB 1600 MHz, 240 GB SSD)

```
Server Software:
Server Hostname:        www.mydomain.com
Server Port:            80

Document Path:          /cameras-camcorders/
Document Length:        58117 bytes

Concurrency Level:      200
Time taken for tests:   6.948 seconds
Complete requests:      10000
Failed requests:        0
Write errors:           0
Total transferred:      582930000 bytes
HTML transferred:       581170000 bytes
Requests per second:    1439.17 [#/sec] (mean)
Time per request:       138.969 [ms] (mean)
Time per request:       0.695 [ms] (mean, across all concurrent requests)
Transfer rate:          81927.36 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    5  16.5      1     256
Processing:     0  134 157.7     70    1028
Waiting:        0   80 124.5     33     987
Total:          0  138 160.5     76    1030

Percentage of the requests served within a certain time (ms)
  50%     76
  66%    131
  75%    190
  80%    232
  90%    361
  95%    477
  98%    692
  99%    732
 100%   1030 (longest request)
```

Enabling again those two cache options from Turpentine is ruining everything. 20 - 30 req/sec, CPU, memory, swap all are blowing up. I have to reboot the system to get it working.

**Comment by ADDISON74** [comment]
```
root@server:~# ab -n10000 -c200 http://www.mydomain.com/cameras-camcorders/
```
Results:

```
This is ApacheBench, Version 2.3 <$Revision: 655654 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking www.mydomain.com (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:
Server Hostname:        www.mydomain.com
Server Port:            80

Document Path:          /cameras-camcorders/
Document Length:        58117 bytes

Concurrency Level:      200
Time taken for tests:   6.477 seconds
Complete requests:      10000
Failed requests:        0
Write errors:           0
Total transferred:      582940000 bytes
HTML transferred:       581170000 bytes
Requests per second:    1543.81 [#/sec] (mean)
Time per request:       129.549 [ms] (mean)
Time per request:       0.648 [ms] (mean, across all concurrent requests)
Transfer rate:          87885.78 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    9  24.1      3     414
Processing:     0  120 140.7     63    1178
Waiting:        0   65 114.9     21    1089
Total:          1  129 144.2     70    1181

Percentage of the requests served within a certain time (ms)
  50%     70
  66%    114
  75%    148
  80%    183
  90%    343
  95%    485
  98%    550
  99%    640
 100%   1181 (longest request)
```
This is a new benchmark with all caches disabled. Before doing the test I deleted Apache access and error starting them from scratch. After the test is finishing checking Apache logs they are untouched. Perfectly normal when everything is server from Varnish and not touching Apache.

**Comment by pawankukreja01** [comment]
The error message you are seeing indicates that the AB test is touching directly the Apache server, which is causing the memory to increase and the system to slow down 
Turpentine is a Varnish cache extension for Magento that is designed to detect AB requests and serve them 1. However, it seems that Turpentine is not detecting the AB requests in your case 1.
There could be several reasons why Turpentine is not detecting the AB requests, such as changes in how cookies are processed or incorrect configuration settings. 
You can try checking the following:
* Check if the cookies are set correctly for the frontend.
* Check if the Turpentine configuration settings are correct.
* Check if there are any changes in how cookies are processed in your system.

**Comment by vinay-agens** [comment]
**Cookie Processing**: By default, Varnish doesn't cache requests with cookies and Magento sends the frontend cookie with every request causing a (near) zero hit rate for Varnish's cache. If there have been changes in how cookies are processed, this could potentially affect the performance.

**Varnish Configuration**: Check your Varnish configuration to ensure it's optimized for your specific use case. You might also want to look at the Turpentine configuration in Magento.

**Server Load**: If your server is under heavy load, this could potentially affect the performance of your ab tests.

**Software Version**: If you've recently upgraded to a new version of Magento, Varnish, or another piece of software, there could be changes or bugs that are affecting performance.
